### PR TITLE
[MM-15368] Support installation version changes

### PR DESCRIPTION
**Pull request by gabrieljackson**
This change introduces support for spinning up installations with
specified versions. It also allows for changing the version of a
previously-running installation.

Fixes https://mattermost.atlassian.net/browse/MM-15368

### CODE REVIEW

File: internal/api/installation_test.go [hunk 1]
```diff
@@ -468,15 +468,6 @@ func TestUpgradeInstallation(t *testing.T) {
 		require.Equal(t, model.InstallationStateUpgradeRequested, installation1.State)
 	})
 
-	t.Run("while stable, to invalid version", func(t *testing.T) {
```
Comment (lieut-data) [inline internal/api/installation_test.go]: Can we replace this test with one that asserts the requested version is written to the installation record?

File: internal/provisioner/kops_provisioner.go [hunk 2]
```diff
@@ -436,8 +436,13 @@ func (provisioner *KopsProvisioner) CreateClusterInstallation(cluster *model.Clu
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      makeClusterInstallationName(clusterInstallation),
 			Namespace: clusterInstallation.Namespace,
+			Labels: map[string]string{
```
Comment (lieut-data) [inline internal/provisioner/kops_provisioner.go]: 🎉

File: internal/model/installation.go [hunk 3]
```diff
@@ -6,23 +6,25 @@ import (
 )
 
 const (
-	// InstallationStateCreationRequested is a installation in the process of being created.
```
Comment (lieut-data) [inline internal/model/installation.go]: 🤦‍♂

File: internal/provisioner/kops_provisioner.go [hunk 4]
```diff
@@ -503,3 +508,119 @@ func (provisioner *KopsProvisioner) DeleteClusterInstallation(cluster *model.Clu
...
+	cr, err := k8sClient.MattermostClientset.MattermostV1alpha1().ClusterInstallations(clusterInstallation.Namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get cluster installation %s", clusterInstallation.ID)
+	}
+
+	logger.WithField("status", fmt.Sprintf("%+v", cr.Status)).Debug("Got cluster installation")
+
+	version := translateMattermostVersion(installation.Version)
+	if cr.Spec.Version == version {
+		return nil
```
Comment (lieut-data) [inline internal/provisioner/kops_provisioner.go]: Should we log something here that we decided not to take any action?

File: internal/supervisor/cluster_installation.go [hunk 5]
```diff
@@ -159,13 +163,28 @@ func (s *ClusterInstallationSupervisor) transitionClusterInstallation(clusterIns
 
 		err = s.store.DeleteClusterInstallation(clusterInstallation.ID)
 		if err != nil {
-			logger.WithError(err).Error("Failed to record deleted cluster installation after creation")
+			logger.WithError(err).Error("Failed to record deleted cluster installation after deletion")
```
Comment (lieut-data) [inline internal/supervisor/cluster_installation.go]: 👍

**Comment by gabrieljackson** [comment]
Thanks for the review @lieut-data!

You are correct that any arbitrary version can be used with respect to the idea that at the end of the day operator is going to look for a docker image of `IMAGE_NAME:VERSION`. This presents a lot of flexibility, but also some challenges. The primary challenge I see is that we really can't do any sanitization of version number to see if it is correct or not unless we want to limit the docker image tags we support.

@cpanato explained to me recently that all PRs get a docker image built and pushed with the tag being the PR number. (apologies if I misunderstood that, but I believe that is what he said). This will work well for spinmints and other test clusters I believe, but also means we can't really limit the semantics of the version number at this time.

All of this may need to be altered in the future, but I believe the implementation I have in this PR will at least let us continue to develop and see how it goes before making further adjustments.

File: internal/model/installation.go (hunk 3 above)
Comment (gabrieljackson) [reply internal/model/installation.go]: Not a big deal! I am sure it was just a minor copy+pasta error.

File: internal/provisioner/kops_provisioner.go (hunk 4 above)
Comment (gabrieljackson) [reply internal/provisioner/kops_provisioner.go]: Sure!

File: internal/api/installation_test.go (hunk 1 above)
Comment (gabrieljackson) [reply internal/api/installation_test.go]: Sure thing!

**Pull request merged**